### PR TITLE
cmake: Move all dependencies to separate targets

### PR DESCRIPTION
**Pull request by jtojnar**
Passing compiler and linker flags manually is not recommended practice any more. Instead, every dependency is supposed to be a separate build target (only marked as imported) that will be imported using find_package and will pass appropriate flags as needed.

https://pabloariasal.github.io/2018/02/19/its-time-to-do-cmake-right/

In particular, this meant creating CMake modules for dependencies whose modules are not shipped with CMake, do not provide their own CMake modules, or the modules do not export targets yet.

It seems to me G'Mic does not actually depend on much other than CImg and that pulls all the dependencies, so I moved the feature flags to the newly created CImg.cmake module.

I also switched to CMake's LTO module (called IPO) since we no longer pass linker flags to G'Mic targets.

Finally, I simplified the CLI target to only depend on the appropriate libgmic.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -319,56 +160,30 @@ endif()
...
-      ${OPENEXR_LIBRARIES}
-      ${OPENCV_LIBRARIES}
-      ${ZLIB_LIBRARIES}
-      ${CURL_LIBRARIES}
-      ${FFTW3_LIBRARIES}
-      ${EXTRA_LIBRARIES}
-    )
+    target_link_libraries(gmic libgmicstatic)
   endif()
+  # TODO: This should probably be inherited from the library by making the include dirs public.
```
Comment (jtojnar) [inline CMakeLists.txt:184]: CMake rightfully does not like if `PUBLIC` include directories point to the source directory so until we figure out how to do it properly we have to use this hack.

Comment (jtojnar) [reply CMakeLists.txt]: The error was:

```
CMake Error in CMakeLists.txt:
  Target "libgmic" INTERFACE_INCLUDE_DIRECTORIES property contains path:

    "/home/jtojnar/Projects/gmic/src"

  which is prefixed in the source directory.

```

Managed to fix it using generator expressions.

**Comment by dtschump** [comment]
Thanks a lot.

As a side note: as you may have noticed, I'm not the author of the CMakeLists.txt file and I have actually no clues on how this whole thing works. All these 'packaging' things bore me to death, and I prefer to focus on the development itself.
I've never used the CMakeList.txt, I've always compiled G'MIC with the custom-made Makefile located in `src/`. That's also the way I compile G'MIC when building the .deb packages provided on the website. To be honest, I've often thought of removing this CMakeLists.txt from the project to avoid having to maintain it (which I don't know how to do). if you're interested in maintaining this particular file, I'd be happy to.

**Comment by jtojnar** [comment]
I do not really like CMake but the current `CMakeLists.txt` are slightly more portable than the `Makefile` in my experience (builds without e.g. curl without having to muck with things like `CURL_FLAGS`, does not rely on global paths like `/usr` to find OpenEXR…). And now that it has been ported to targets, also much cleaner architecture-wise.

Eventually, I would like to add CMake (or maybe Meson, which is nicer to use but maybe not as flexible) to CImg to be able to install it globally instead of picking it from the source zip but will have to research how to make configurable CMake modules first.

http://www.mariobadr.com/creating-a-header-only-library-with-cmake.html
https://pabloariasal.github.io/2018/02/19/its-time-to-do-cmake-right/

**Pull request merged**